core/payment-driver/erc20/src/erc20/ethereum.rs

```diff
@@ -350,69 +478,152 @@ pub async fn decode_encoded_transaction_data(
             return Ok((add, am));
         }
     }
-    Err(GenericError::new("Failed to parse tokens"))
+    return Err(GenericError::new("Failed to parse tokens").into());
 }
 
 pub async fn get_tx_from_network(
     tx_hash: H256,
     network: Network,
 ) -> Result<Option<Transaction>, GenericError> {
-    let client = get_client(network).await?;
-    let result = client
+    with_clients(network, |client| get_tx_from_network_with(tx_hash, client)).await
+}
+
+async fn get_tx_from_network_with(
+    tx_hash: H256,
+    client: Web3<Http>,
+) -> Result<Option<Transaction>, CustomError> {
+    client
         .eth()
         .transaction(TransactionId::from(tx_hash))
         .await
-        .map_err(GenericError::new)?;
-    Ok(result)
+        .map_err(Into::into)
 }
 
 pub async fn get_tx_receipt(
     tx_hash: H256,
     network: Network,
 ) -> Result<Option<TransactionReceipt>, GenericError> {
-    let client = get_client(network).await?;
-    let result = client
+    with_clients(network, |client| get_tx_receipt_with(tx_hash, client)).await
+}
+
+async fn get_tx_receipt_with(
+    tx_hash: H256,
+    client: Web3<Http>,
+) -> Result<Option<TransactionReceipt>, CustomError> {
+    client
         .eth()
         .transaction_receipt(tx_hash)
         .await
-        .map_err(GenericError::new)?;
-    Ok(result)
+        .map_err(Into::into)
 }
 
-fn get_rpc_addr_from_env(network: Network) -> String {
+fn get_rpc_addr_from_env(network: Network) -> Vec<String> {
     match network {
-        Network::Mainnet => std::env::var("MAINNET_GETH_ADDR")
-            .unwrap_or("https://geth.golem.network:55555".to_string()),
-        Network::Rinkeby => std::env::var("RINKEBY_GETH_ADDR")
-            .unwrap_or("http://geth.testnet.golem.network:55555".to_string()),
+        Network::Mainnet => {
+            let mut vec: Vec<String> = Default::default();
+            let env = std::env::var("MAINNET_GETH_ADDR").ok();
+            if let Some(env) = env {
+                vec.push(env)
+            }
+
+            for (key, value) in std::env::vars() {
+                if key.contains("MAINNET_GETH_ADDR_") {
+                    vec.push(value)
+                }
+            }
+            vec.push("https://geth.golem.network:55555".to_string());
+            vec
```

Comment:
Could `MAINNET_GETH_ADDR` and other similar values contain e.g. a comma-separated list of addresses? Why are separate env variables needed?